### ISSUE TITLE
Config.ClearType description is unclear

### ISSUE THREAD

**Issue by nargil** [kolton/d2bot-with-kolbot#2040]
// Monster spectype to kill in level clear scripts (ie. Mausoleum). 0xF = skip normal, 0x7 = champions/bosses, 0 = all

So does it actually kill more monsters (monster types) with 0xF or 0x7 ?

**Comment by anime143** [comment]
put 0, this will do both

**Comment by nargil** [comment]
No, that's not the answer to the question. (perhaps you don't know either?)
I had to look it up in the source code. If you treat it as a bitmask filter:
0x0 = 00000000
0x7 = 00001111
0xF = 11111111

and you know 0x0 kills everythink, you come up to the conclusion that this is an exclusion filter, so 0x7 would kill more than 0xF.

However this is not the case, as the code is 
if ((!spectype || (target.spectype & spectype)) ...) {
   ...
   monsterList.push(copyUnit(target));
}

so in fact it's an inclusion filter, so 0xF includes more monsters to kill than 0x7, and the fact that 0x0 kills more than 0xF is due to the "or nullity check".

I really think this should be better documented.

**Comment by tlwh163** [comment]
0x7 = 00000111
0xF = 00001111

**Comment by bluedevilsteve** [comment]
i've always thought it was not commented well. should be clearer in config files.

**Comment by Fa-b** [comment]
Hmm.. I did some research on this. 'spectype' is not coming from kolbot itself but from d2bs.

Check API here: [https://github.com/noah-/d2bs#classUnit](https://github.com/noah-/d2bs#classUnit)

According to this, spectype differentiates between 4 types of monsters:
1 - "Normal" Boss
2 - Champion
4 - Boss
8 - Minion

unfortunately JavaScript is lacking a real enumeration type, otherwise many of these things would be more selfexplanatory...
@nargil is right that kolbot is using an inclusion filter. Without the check for !spectype though, the bot would probably never kill normal monsters since 'spectype' only refers to 'special type' and `(target.spectype & spectype)` would not be true for the value of <b>null</b>...

basically you can specify any combination of monster types to clear, but normal monsters will always have to be skipped if any other value then 0 is passed:

| spectype | monster types to kill |
| ------------- | ------------- |
| 0x0 | all |
| 0x1 | "Normal" Boss |
| 0x2 | Champion |
| 0x3 | "Normal" Boss + Champion |
| 0x4 | Boss |
| 0x5 | "Normal" Boss + Boss |
| 0x6 | Champion + Boss |
| 0x7 | "Normal" Boss + Champion + Boss |
| 0x8 | Minion |
| 0x9 | "Normal" Boss + Minion |
| 0xA | Champion + Minion |
| 0xB | "Normal" Boss + Champion + Minion |
| 0xC | Boss + Minion |
| 0xD | "Normal" Boss + Boss + Minion |
| 0xE | Champion + Boss + Minion |
| 0xF |  "Normal" Boss + Champion + Boss + Minion |

I don't know if that nomenclature helps though.. or if there is a difference between 'Minions' and 'Normal monsters'.

Maybe someone can add to this
Fa-b

**Comment by getmewings** [comment]
Not sure if this was determined already above but the difference would be 0xF will kill the boss, say for example rakanishu, AND all his little friends that hang out near the Tristram portal. 0x7 would simply kill rakanishu and then move on whether or not the minions are killed.
